vosk/src/recognition/mod.rs

```diff
@@ -277,3 +283,86 @@ impl Drop for Recognizer {
+    use crate::batch_model::BatchModel;
+    use vosk_sys::*;
+
+    use std::{ffi::CStr, ptr::NonNull};
+
+    pub use crate::recognition::results::*;
+
+    /// The main object which processes data using GPU inferencing.
+    /// Takes audio as input and returns decoded information as words, confidences, times, and other metadata.
+
+    pub struct BatchRecognizer(std::ptr::NonNull<VoskBatchRecognizer>);
+
+    impl BatchRecognizer {
+        /// Creates the recognizer object. Returns [`None`] if a problem occured.
+        ///
+        /// The recognizers process the speech and return text using shared model data.
+        ///
+        /// * `model` - [`BatchModel`] containing static data for recognizer. Model can be shared
+        /// across recognizers, even running in different threads.
+        ///
+        /// * `sample_rate` - The sample rate of the audio you going to feed into the recognizer.
+        /// Make sure this rate matches the audio content, it is a common issue causing accuracy problems.
+        ///
+        /// [`BatchModel`]: crate::BatchModel
+        #[must_use]
+        pub fn new(model: &BatchModel, sample_rate: f32) -> Option<Self> {
+            let recognizer_ptr =
+                unsafe { vosk_batch_recognizer_new(model.0.as_ptr(), sample_rate) };
+            Some(Self(NonNull::new(recognizer_ptr)?))
+        }
+
+        /// Enables or disables Natural Language Semantics Markup Language (NLSML) in the output
+        pub fn set_nlsml(&mut self, enable: bool) {
+            unsafe { vosk_batch_recognizer_set_nlsml(self.0.as_ptr(), i32::from(enable)) }
+        }
+
+        /// Accept and process new chunk of voice data.
+        ///
+        /// * `data` - Audio data in PCM 16-bit mono format as an array of i8.
+        pub fn accept_waveform(&mut self, data: &[i8]) {
+            unsafe {
+                vosk_batch_recognizer_accept_waveform(
+                    self.0.as_ptr(),
+                    data.as_ptr(),
+                    data.len() as i32,
+                )
+            };
+        }
+
+        /// Closes the stream to the model
+        pub fn finish_stream(&mut self) {
+            unsafe { vosk_batch_recognizer_finish_stream(self.0.as_ptr()) };
+        }
+
+        /// Gets the front of the result queue
+        pub fn front_result(&mut self) -> Result<Word, serde_json::Error> {
+            serde_json::from_str(
+                unsafe { CStr::from_ptr(vosk_batch_recognizer_front_result(self.0.as_ptr())) }
+                    .to_str()
+                    .unwrap(),
+            )
+        }
+
+        /// Removes the front of the result queue
+        pub fn pop(&mut self) {
+            unsafe { vosk_batch_recognizer_pop(self.0.as_ptr()) }
+        }
+
+        /// Gets the number of chunks that have yet to be processed
+        pub fn get_pending_chunks(&mut self) -> usize {
+            (unsafe { vosk_batch_recognizer_get_pending_chunks(self.0.as_ptr()) }) as usize
```

Comment:
I don't think this should return `usize` as it'd be potentially wasting memory. The real number will only have the range of a `c_int` (`i32`) as per the underliying type in Vosk.

I suggest the following: correct me if I'm wrong, but I am pretty sure the chunk count will only be `>= 0` since it doesn't make sense to have "a negative of chunks left" and I checked the underlying code and it doesn't seem to use negative numbers for error codes. If that's the case, we could change the return type of this function to a `u16` to better describe the real range of the returned values.

Reply:
Using a `u16` would probably be fine, but all the `CudaOnlinePipelineDynamicBatcher` stuff is pretty esoteric. 16 bits isn't that much memory to use just to be sure, but if you feel really strongly about it, I (or you) can change it to `u16`.

Reply:
Nah it was just a suggestion. I have barely no background on GPU programming so I prefer to follow your advice. Plus in a way it's also better to just leave it as the original type since down the road it could prevent issues caused by unforseen values returned by vosk. This way is just simpler overall ;)